File: apps/client-macos/Sources/TimeTrack/App/MenuViewModel.swift
import Foundation
import SwiftUI

/// One flattened project/task the picker can select. `taskId == nil` is the project itself.
struct Choice: Identifiable, Equatable {
    let id: String        // taskId ?? projectId
    let projectId: String
    let taskId: String?
    let projectName: String
    let taskName: String?
}

/// The single seam between the SwiftUI dropdown and the tracking logic. The view observes
/// this; the view has no logic and `TimeTracker` has no UI. Manual tracking is gated here by
/// `isReady` (set by AppDelegate once the launch ack flow resolves) — NOT by AckGate
/// (CLAUDE.md §1); AckGate stays reserved for real capture in 1.7c+.
///
/// A plain ObservableObject (not @MainActor): every @Published mutation is driven from the
/// main thread — SwiftUI button actions or AppDelegate's `MainActor.run` (the 1.7a pattern).
final class MenuViewModel: ObservableObject {
    enum Phase: Equatable { case idle, tracking, paused }

    @Published private(set) var isReady = false
    /// Whether a user session is open. Drives the dropdown's signed-in vs signed-out layout —
    /// the signed-out dropdown never offers My Data / Sign Out (those belong to a session).
    @Published private(set) var isSignedIn = false
    @Published private(set) var phase: Phase = .idle
    @Published private(set) var startedAt: Date?
    @Published private(set) var selectedChoice: Choice?
    @Published var projects: [Project] = []
    @Published var query: String = ""

    /// What the person says they were doing. Typed at any point during a span and applied to
    /// the RUNNING entry in place — a note does not re-attribute time the way a project switch
    /// does, so it must not split the span at the moment they stopped typing.
    ///
    /// Cleared on stop and on sign-out, so the next entry never inherits the last one's note.
    @Published var note: String = "" {
        didSet { tracker.setNote(trimmedNote) }
    }

    private var trimmedNote: String? {
        let t = note.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
    }

    /// The server's own today / this-week / this-month totals, refreshed when the dropdown opens.
    ///
    /// Nil means "not known yet" and renders as an em dash — never as 0h, which would be a lie
    /// about someone's tracked time on the one screen where that number is the whole point. A
    /// refresh that FAILS leaves the last successful value in place: it was true when it was
    /// fetched, and that is more useful than blanking the row on a dropped connection.
    ///
    /// These tick. The server's sum already counts a running entry up to the moment of the
    /// fetch — while heartbeats are fresh the clamp resolves to `now()` — so the view adds only
    /// the wall time elapsed SINCE the fetch, and today's figure moves as the person works.
    /// See `totalsFetchedAt`.
    @Published var totals: SelfTotals?

    /// When `totals` was fetched. This is what makes them live rather than a snapshot: the base
    /// number was true at this instant, so the running session's contribution since then is
    /// exactly `now - max(this, startedAt)`. Anchoring on the fetch instead of the session start
    /// is what keeps the current session from being counted twice — the base already contains it
    /// up to here.
    @Published var totalsFetchedAt: Date?

    /// How many of this person's own records have not reached the server yet — closed time
    /// entries, idle events, and screenshots still sitting in the durable buffers.
    ///
    /// Distinct from the menu bar's "not recording" warning, which says the server is actively
    /// REJECTING the running entry. This one is the ordinary offline case: nothing is wrong,
    /// the work is safely on disk, and it will drain. Showing it means a person who has been on
    /// a plane can see their time is not lost instead of guessing.
    @Published var pendingSyncCount: Int = 0

    /// The signed-in user, set by AppDelegate once the session resolves. Selection persistence is
    /// namespaced by it so one user can never inherit another's (possibly wrong-team) project.
    var currentUserId: String?

    /// Called on every tracking-state change with `iconIsTracking` and the tracking start (nil
    /// when not tracking), so AppDelegate can update the always-visible status icon — including
    /// its live elapsed count — without a Combine subscription.
    var onPhaseChanged: ((Bool, Date?) -> Void)?

    private let tracker: TimeTracker
    private let dashboardURL: URL
    private let openURL: (URL) -> Void
    private let onSignIn: () -> Void
    private let onSignOut: () -> Void
    private let onQuit: () -> Void
    private let selectionStore: SelectionStore

    /// A display-only clock anchor that overrides the live entry's real start. Set when manual idle
    /// Discard shifts the clock forward by the idle gap so it keeps reading accumulated *worked*
    /// time (the fresh entry's real start would read 0). Display-only: the entry sent to the server
    /// keeps its true start. Cleared by any explicit start/stop/pause/resume/reset, so a new user
    /// session reads its real elapsed again.
    private var displayStartOverride: Date?

    init(
        tracker: TimeTracker,
        dashboardURL: URL,
        openURL: @escaping (URL) -> Void,
        onSignIn: @escaping () -> Void,
        onSignOut: @escaping () -> Void,
        onQuit: @escaping () -> Void,
        selectionStore: SelectionStore
    ) {
        self.tracker = tracker
        self.dashboardURL = dashboardURL
        self.openURL = openURL
        self.onSignIn = onSignIn
        self.onSignOut = onSignOut
        self.onQuit = onQuit
        self.selectionStore = selectionStore
    }

    /// The indicator reads as active whenever a session is open (tracking or paused) — it is
    /// never hidden while a span is in progress (PRD §4.2). Mapping `paused` → the tracking
    /// icon is deliberate over-disclosure: for monitoring software the indicator errs toward
    /// "active", never toward under-reporting. (`StatusItemController.State` has no `paused`.)
    var iconIsTracking: Bool { phase != .idle }

    /// The current picker selection as a tracker Selection, for auto-started entries
    /// (they inherit whatever the employee has picked; null if nothing is selected).
    var selectionForAuto: TimeTracker.Selection {
        TimeTracker.Selection(projectId: selectedChoice?.projectId, taskId: selectedChoice?.taskId)
    }

    var choices: [Choice] {
        projects.flatMap { p -> [Choice] in
            let projectOnly = Choice(id: p.id, projectId: p.id, taskId: nil,
                                     projectName: p.name, taskName: nil)
            let tasks = (p.tasks ?? []).map {
                Choice(id: $0.id, projectId: p.id, taskId: $0.id,
                       projectName: p.name, taskName: $0.name)
            }
            return [projectOnly] + tasks
        }
    }

    var filteredChoices: [Choice] {
        guard !query.isEmpty else { return choices }
        let q = query.lowercased()
        return choices.filter {
            $0.projectName.lowercased().contains(q) || ($0.taskName?.lowercased().contains(q) ?? false)
        }
    }

    func markReady() {
        isReady = true
        isSignedIn = true
    }
    func markNotReady() { isReady = false }

    /// Pick a project/task. The menu calls this "SWITCH PROJECT", and it means it: if a span is
    /// already running, the running entry is CLOSED here and a fresh one opened under the new
    /// selection.
    ///
    /// Storing the choice alone was not enough. `TimeTracker` captures the selection when a span
    /// OPENS and enqueues that captured value when it closes, so a switch made mid-span left the
    /// whole span — everything before the switch AND everything after — filed under the previous
    /// project. In AUTO mode a span runs from login to the first idle window, so that is hours of
    /// work attributed to the wrong project, with the UI reporting the switch as done.
    ///
    /// A `.paused` session is re-armed the same way: `resume()` reopens from the selection
    /// captured in the paused state, so without this a switch while paused would also be lost.
    ///
    /// The display clock is anchored to the OLD start, so the header keeps reading accumulated
    /// worked time instead of snapping back to 0 — the same treatment the manual-idle Discard
    /// trim gets. The entry itself starts now; only the readout continues.
    func select(_ choice: Choice) {
        selectedChoice = choice
        persist(choice)

        switch tracker.state {
        case let .tracking(_, entryStart, _, source):
            let anchor = displayStartOverride ?? entryStart
            tracker.stop()
            tracker.start(projectId: choice.projectId, taskId: choice.taskId,
                          note: trimmedNote, source: source)
            displayStartOverride = anchor
            sync()
        case .paused:
            // Nothing is running, so nothing to re-file — but the paused selection is what
            // `resume()` reopens with, so it has to be replaced rather than remembered.
            tracker.pause(reselecting: TimeTracker.Selection(projectId: choice.projectId,
                                                             taskId: choice.taskId,
                                                             note: trimmedNote))
        case .idle:
            break
        }
    }

    private func persist(_ choice: Choice) {
        guard let currentUserId else { return }
        selectionStore.save(StoredSelection(projectId: choice.projectId, taskId: choice.taskId),
                            userId: currentUserId)
    }

    /// Apply the persisted selection for `userId`, if it still exists in the CURRENT project
    /// list. Called by AppDelegate AFTER the project refresh — restoring against a stale or
    /// empty cache would either drop a valid selection or apply one the user has since lost
    /// access to.
    ///
    /// Only ever fills an EMPTY picker (`selectedChoice == nil`): `refreshProjects()` runs both
    /// at launch and on every menu open, so an unguarded restore would re-run each time and could
    /// overwrite a selection the user just made by hand.
    ///
    /// A selection that no longer resolves is dropped, but the stored key is cleared only when
    /// there is a non-empty project list to judge it against. `ProjectCache` is a single global
    /// file cleared wholesale on sign-out, so `projects` reads empty on every offline re-login
    /// until a network refresh succeeds — indistinguishable, to `SelectionResolver`, from "the
    /// project was archived". Clearing on an empty list would permanently delete a user's saved
    /// selection on an offline re-login instead of merely deferring the restore.
    func restoreSelection(userId: String) {
        guard selectedChoice == nil else { return }
        guard let stored = selectionStore.load(userId: userId) else { return }
        if let restored = SelectionResolver.resolve(stored, in: choices) {
            selectedChoice = restored
        } else if !choices.isEmpty {
            selectionStore.clear(userId: userId)
        }
    }

    func start() {
        guard isReady else { return }
        displayStartOverride = nil
        tracker.start(projectId: selectedChoice?.projectId, taskId: selectedChoice?.taskId,
                      note: trimmedNote)
        sync()
    }

    /// Start, or stop, whichever the current state calls for — what the global shortcut does.
    ///
    /// A paused session RESUMES rather than starting fresh, so the shortcut can never silently
    /// abandon a pause the person meant to come back to.
    func toggle() {
        switch tracker.state {
        case .tracking: stop()
        case .paused: resume()
        case .idle: start()
        }
    }

    func stop() {
        displayStartOverride = nil
        tracker.stop() // closes with the note captured on the span
        note = ""
        sync()
    }
    func pause() { displayStartOverride = nil; tracker.pause(); sync() }

    /// Manual idle Discard has replaced the live entry (trim + fresh start). Shift the clock anchor
    /// forward by the discarded idle gap so it keeps reading accumulated *worked* time and climbs,
    /// rather than resetting to the fresh entry's 0. The selection (project/task) is preserved —
    /// the fresh entry inherits it and the picker is untouched. Then re-sync the UI + status icon.
    func continueClockAfterDiscard(idleSeconds: TimeInterval) {
        if let anchor = startedAt {
            displayStartOverride = anchor.addingTimeInterval(idleSeconds)
        }
        sync()
    }

    func resume() {
        guard isReady else { return }
        displayStartOverride = nil
        tracker.resume()
        sync()
    }

    /// Open the employee self-view directly (the dashboard's /me page), not the dashboard root —
    /// the rich "My data" screen (time ribbon, activity, screenshots + redaction) lives on the web.
    func openMyData() { openURL(dashboardURL.appendingPathComponent("me")) }

    /// Returns the VM to a clean signed-out state: closes/enqueues any live span (`stop()`
    /// drives phase→idle + icon via `sync()`) and clears everything selection/search/project
    /// related so a different user's next login can't inherit a stale, wrong-team selection
    /// (CLAUDE.md §1 fail-safe posture).
    ///
    /// Deliberately asymmetric: this clears the IN-MEMORY selection and drops `currentUserId`
    /// (so a stray `select()` before the next sign-in can never write under the old user), but
    /// does NOT clear the PERSISTED key — the per-user namespacing is what lets the same user get
    /// their project back on their next sign-in without reopening the cross-user leak this guards.
    func reset() {
        stop()
        isReady = false
        isSignedIn = false
        selectedChoice = nil
        query = ""
        projects = []
        currentUserId = nil
        // `stop()` above already cleared it; repeated here because the note is user-authored
        // text and this is the cross-user boundary — the same class of leak that has bitten
        // the away/recovery prompts twice. A reorder of this method must not reopen it.
        note = ""

        // A previous user's tracked time must never be visible to whoever signs in next.
        totals = nil
        totalsFetchedAt = nil
        pendingSyncCount = 0
    }

    /// A total as it stands right now: the figure the server returned, plus the tracked time that
    /// has accrued since it was fetched.
    ///
    /// Anchored on `max(fetchedAt, startedAt)`, which is the whole trick:
    ///  * the base ALREADY includes the running session up to `fetchedAt`, so counting from the
    ///    session start would count most of it twice;
    ///  * a session that began AFTER the fetch has only run since `startedAt`, so counting from
    ///    the fetch would add idle time that was never tracked.
    ///
    /// Only while the clock is actually running — a paused or stopped clock accrues nothing.
    static func liveTotal(base: Int, phase: Phase, fetchedAt: Date?, startedAt: Date?, now: Date) -> Int {
        guard phase == .tracking, let fetchedAt, let startedAt else { return base }
        let since = max(fetchedAt, startedAt)
        return base + max(0, Int(now.timeIntervalSince(since)))
    }

    /// Bring the sign-in window forward from the signed-out dropdown (the login window is
    /// already presented on sign-out, but the user may have dismissed it).
    func signIn() { onSignIn() }

    func signOut() {
        reset()
        onSignOut()
    }

    func quit() { onQuit() }

    private func sync() {
        switch tracker.state {
        case .idle:
            phase = .idle
            startedAt = nil
            displayStartOverride = nil
        case let .tracking(_, entryStart, _, _):
            phase = .tracking
            // Prefer the Discard clock anchor (worked-time continuation) over the entry's real
            // start; the override is display-only and never leaves this type.
            startedAt = displayStartOverride ?? entryStart
        case .paused:
            phase = .paused
            startedAt = nil
        }
        onPhaseChanged?(iconIsTracking, startedAt)
    }
}

File: apps/client-windows/src/NiftyTimer/App/MenuViewModel.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using NiftyTimer.Projects;
using NiftyTimer.Reports;
using NiftyTimer.Tracking;

namespace NiftyTimer.App;

/// <summary>One row in the project picker: a project on its own, or one of its tasks.</summary>
public sealed record PickerChoice(string ProjectId, string? TaskId, string ProjectName, string? TaskName);

/// <summary>
/// What the tray dropdown shows and what it can do. UI-thread-only.
///
/// This is where readiness is enforced for MANUAL tracking: <see cref="TimeTracker"/> is not
/// behind <see cref="Policy.AckGate"/> (it is not a capture path — CLAUDE.md §1), so the rule
/// that an un-acknowledged user cannot start the clock lives here, in <see cref="IsReady"/>.
/// Capture paths do NOT rely on this; they are gated structurally and are installed only on the
/// online, acknowledged branch.
/// </summary>
public sealed class MenuViewModel : INotifyPropertyChanged
{
    private readonly TimeTracker _tracker;
    private readonly SelectionStore _selectionStore;
    private readonly Func<DateTimeOffset> _clock;

    private bool _isReady;
    private string? _userId;
    private IReadOnlyList<Project> _projects = [];
    private StoredSelection? _selection;
    private SelfTotals? _totals;
    private int _pendingCount;
    private bool _liveSyncBlocked;
    private bool _updateAvailable;
    private bool _updateOverdue;
    private string? _notice;
    private string _note = string.Empty;
    private string _query = string.Empty;
    private DateTimeOffset? _displayStart;
    private DateTimeOffset? _totalsFetchedAt;
    private bool _wasTracking;

    public MenuViewModel(
        TimeTracker tracker,
        SelectionStore selectionStore,
        Func<DateTimeOffset>? clock = null)
    {
        _tracker = tracker;
        _selectionStore = selectionStore;
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    /// <summary>Raised when the user asks to start tracking and the view model allowed it.</summary>
    public event Action? TrackingStarted;

    /// <summary>
    /// The clock went from running to not running — stopped, paused, rolled back, or auto-stopped.
    /// The live increment on the totals ends at that moment, so the figures would drop back to the
    /// last fetch; this is the cue to fetch fresh ones rather than show time going backwards.
    /// </summary>
    public event Action? TrackingStopped;

    /// <summary>
    /// True once the session is usable for manual tracking: signed in, and the monitoring policy
    /// acknowledged (either confirmed online this launch, or recorded locally by
    /// <see cref="Policy.AckMarker"/> on a previous one).
    /// </summary>
    public bool IsReady
    {
        get => _isReady;
        set => Set(ref _isReady, value, [nameof(CanStart), nameof(CanStop)]);
    }

    public string? UserId
    {
        get => _userId;
        set => Set(ref _userId, value);
    }

    public IReadOnlyList<Project> Projects
    {
        get => _projects;
        set => Set(ref _projects, value, [nameof(Choices), nameof(FilteredChoices)]);
    }

    /// <summary>
    /// What the person has typed into the picker's search field. Kept here rather than in the
    /// popup so it survives the popup hiding and showing, as the macOS dropdown's does, and so the
    /// filter is testable without a window.
    /// </summary>
    public string Query
    {
        get => _query;
        set => Set(ref _query, value ?? string.Empty, [nameof(FilteredChoices)]);
    }

    /// <summary>Every row the picker can offer: each project, then each of its tasks.</summary>
    public IReadOnlyList<PickerChoice> Choices => ChoicesFor(_projects);

    /// <summary>The rows matching <see cref="Query"/>. See <see cref="Filter"/>.</summary>
    public IReadOnlyList<PickerChoice> FilteredChoices => Filter(Choices, _query);

    /// <summary>
    /// A row matches when the query appears ANYWHERE in its project or task name, ignoring case —
    /// the macOS client's rule. The combo box this replaced only matched a prefix of the whole
    /// label, so "design" could not find "Website · Design review" at all.
    /// </summary>
    public static IReadOnlyList<PickerChoice> Filter(IReadOnlyList<PickerChoice> choices, string query)
    {
        if (string.IsNullOrEmpty(query))
        {
            return choices;
        }

        return choices
            .Where(c => c.ProjectName.Contains(query, StringComparison.OrdinalIgnoreCase)
                        || (c.TaskName?.Contains(query, StringComparison.OrdinalIgnoreCase) ?? false))
            .ToList();
    }

    private static List<PickerChoice> ChoicesFor(IReadOnlyList<Project> projects)
    {
        var choices = new List<PickerChoice>();
        foreach (var project in projects)
        {
            choices.Add(new PickerChoice(project.Id, null, project.Name, null));
            foreach (var task in project.Tasks ?? [])
            {
                choices.Add(new PickerChoice(project.Id, task.Id, project.Name, task.Name));
            }
        }

        return choices;
    }

    public StoredSelection? Selection
    {
        get => _selection;
        private set => Set(ref _selection, value, [nameof(SelectionLabel)]);
    }

    /// <summary>
    /// The server's figures, as of the moment they were assigned. That moment is what makes them
    /// live rather than a snapshot — see <see cref="LiveTotal"/>.
    /// </summary>
    public SelfTotals? Totals
    {
        get => _totals;
        set
        {
            // Stamped on every arrival, even an identical one: the base was true at THIS instant,
            // and the live increment counts from here.
            _totalsFetchedAt = value is null ? null : _clock();
            if (!Set(ref _totals, value, [nameof(TodayLabel), nameof(WeekLabel), nameof(MonthLabel)]))
            {
                Raise(nameof(TodayLabel), nameof(WeekLabel), nameof(MonthLabel));
            }
        }
    }

    /// <summary>How many records are still waiting to reach the server.</summary>
    public int PendingCount
    {
        get => _pendingCount;
        set => Set(ref _pendingCount, value, [nameof(PendingLabel), nameof(HasPending)]);
    }

    /// <summary>
    /// The running entry has failed to reach the server several times running. Surfaced because
    /// the alternative — a clock that looks like it is recording while the server hears nothing —
    /// is invisible from both ends.
    /// </summary>
    public bool LiveSyncBlocked
    {
        get => _liveSyncBlocked;
        set => Set(ref _liveSyncBlocked, value);
    }

    /// <summary>
    /// A newer build has been published. Advisory only — nothing in the update path may stop,
    /// block or alter tracking, so this drives a menu line and nothing else.
    /// </summary>
    public bool UpdateAvailable
    {
        get => _updateAvailable;
        set => Set(ref _updateAvailable, value);
    }

    /// <summary>The newer build has been available past the grace period. Adds a tray marker.</summary>
    public bool UpdateOverdue
    {
        get => _updateOverdue;
        set => Set(ref _updateOverdue, value);
    }

    /// <summary>A one-line message for the user; null when there is nothing to say.</summary>
    public string? Notice
    {
        get => _notice;
        set => Set(ref _notice, value, [nameof(HasNotice)]);
    }

    public bool HasNotice => !string.IsNullOrEmpty(_notice);

    /// <summary>The free-text note applied to the running span, in place.</summary>
    public string Note
    {
        get => _note;
        set
        {
            if (Set(ref _note, value))
            {
                _tracker.SetNote(string.IsNullOrWhiteSpace(value) ? null : value);
            }
        }
    }

    public bool IsTracking => _tracker.IsRunning;

    public bool IsPaused => _tracker.IsPaused;

    public bool CanStart => IsReady && !IsTracking;

    public bool CanStop => IsReady && (IsTracking || IsPaused);

    public bool HasPending => _pendingCount > 0;

    public string PendingLabel => _pendingCount == 1 ? "1 record pending" : $"{_pendingCount} records pending";

    public string ElapsedLabel => WorkTotalFormat.Elapsed(Elapsed);

    /// <summary>
    /// How long the dropdown says you have been working on this stretch.
    ///
    /// Usually the running entry's own age. The exception is a discarded idle window: that trims
    /// the entry and opens a fresh one, so the entry's age restarts at zero even though the person
    /// worked for an hour before stepping away. <see cref="ContinueClockAfterDiscard"/> supplies the
    /// instant to count from instead, so the clock keeps reading accumulated WORKED time.
    /// </summary>
    public TimeSpan Elapsed =>
        _tracker.State is TrackerState.Tracking t ? _clock() - (_displayStart ?? t.StartedAt) : TimeSpan.Zero;

    /// <summary>
    /// The selection an auto-started span should carry. Never a note: a note is something the
    /// person typed about work they chose to record, and attaching it to time the machine started
    /// on their behalf would put words in their mouth.
    /// </summary>
    public TimeTracker.Selection SelectionForAuto => new(_selection?.ProjectId, _selection?.TaskId);

    public string TodayLabel => _totals is null ? "—" : WorkTotalFormat.Short(Live(_totals.TodaySeconds));

    public string WeekLabel => _totals is null ? "—" : WorkTotalFormat.Short(Live(_totals.WeekSeconds));

    public string MonthLabel => _totals is null ? "—" : WorkTotalFormat.Short(Live(_totals.MonthSeconds));

    /// <summary>
    /// A total as it stands right now: the server's figure plus the tracked time accrued since it
    /// was fetched. The same increment applies to all three — a minute worked now is a minute of
    /// today, of this week and of this month.
    ///
    /// Anchored on <c>max(fetchedAt, runningSince)</c>, which is the whole trick:
    /// <list type="bullet">
    ///   <item>the fetched figure ALREADY counts the running session up to the fetch, so counting
    ///   from the session start would count most of it twice;</item>
    ///   <item>a session that began AFTER the fetch has only run since its start, so counting from
    ///   the fetch would add time that was never tracked.</item>
    /// </list>
    /// Only while the clock actually runs — paused or stopped accrues nothing.
    /// </summary>
    public static int LiveTotal(int baseSeconds, DateTimeOffset? runningSince, DateTimeOffset? fetchedAt, DateTimeOffset now)
    {
        if (runningSince is not { } start || fetchedAt is not { } fetched)
        {
            return baseSeconds;
        }

        var since = fetched > start ? fetched : start;
        return baseSeconds + Math.Max(0, (int)(now - since).TotalSeconds);
    }

    // The entry's REAL start, not the display anchor: the server counts the entry itself, so the
    // anchor a discarded idle window leaves behind would double-count the discarded minutes.
    private int Live(int baseSeconds) =>
        LiveTotal(
            baseSeconds,
            _tracker.State is TrackerState.Tracking t ? t.StartedAt : null,
            _totalsFetchedAt,
            _clock());

    public string SelectionLabel
    {
        get
        {
            if (_selection is null)
            {
                return "No project";
            }

            var project = _projects.FirstOrDefault(p => p.Id == _selection.ProjectId);
            if (project is null)
            {
                return "No project";
            }

            var task = _selection.TaskId is null
                ? null
                : project.Tasks?.FirstOrDefault(t => t.Id == _selection.TaskId);

            return task is null ? project.Name : $"{project.Name} · {task.Name}";
        }
    }

    public void Start()
    {
        if (!CanStart)
        {
            return;
        }

        Notice = null;
        _displayStart = null;
        _tracker.Start(_selection?.ProjectId, _selection?.TaskId, NoteOrNull());
        TrackingStarted?.Invoke();
        RaiseTrackingState();
    }

    /// <summary>
    /// What the global hotkey does: resume if paused, stop if running, start if we can, and
    /// otherwise do nothing at all. Silence is the right answer for the last case — the hotkey
    /// fires from whatever application has focus, so a person who pressed it before signing in
    /// should not be interrupted by an error they did not ask for.
    ///
    /// Paused resumes rather than stops, matching the macOS client: a pause is a break the person
    /// means to come back from, and the one-key way back should not end the session instead.
    /// </summary>
    public void ToggleTracking()
    {
        if (IsPaused)
        {
            Resume();
        }
        else if (IsTracking)
        {
            Stop();
        }
        else if (CanStart)
        {
            Start();
        }
    }

    public void Stop()
    {
        if (!CanStop)
        {
            return;
        }

        _displayStart = null;
        _tracker.Stop();
        RaiseTrackingState();
    }

    public void Pause()
    {
        _displayStart = null;
        _tracker.Pause();
        RaiseTrackingState();
    }

    public void Resume()
    {
        if (!IsReady)
        {
            return;
        }

        _displayStart = null;
        _tracker.Resume();
        RaiseTrackingState();
    }

    /// <summary>
    /// Re-read the tracker and republish the derived state.
    ///
    /// The manual affordances all mutate the tracker THROUGH this view model, so they can raise as
    /// they go. The AUTO path does not — <see cref="AutoTrackingCoordinator"/> writes straight to
    /// <see cref="TimeTracker"/> — so an auto open or close is invisible from here until something
    /// asks. This is that seam, and it derives everything from the tracker rather than taking a
    /// state argument, so it cannot drift from what the tracker actually holds.
    /// </summary>
    public void RefreshFromTracker()
    {
        _displayStart = null;
        RaiseTrackingState();
    }

    /// <summary>
    /// A discarded idle window replaced the running entry (see
    /// <see cref="ManualIdleCoordinator"/>). <paramref name="displayStart"/> is the instant the
    /// clock should count from so it keeps reading worked time rather than jumping back to zero —
    /// the swap happens directly on <see cref="TimeTracker"/> and is invisible from here otherwise.
    /// </summary>
    public void ContinueClockAfterDiscard(DateTimeOffset displayStart)
    {
        _displayStart = displayStart;
        RaiseTrackingState();
    }

    /// <summary>
    /// The server refused to open <paramref name="entryId"/>: this user is already tracking on
    /// another machine. The clock is rolled back rather than left running against a span the
    /// server never accepted.
    ///
    /// Ignored if that span is no longer the running one — the 409 answers a fire-and-forget
    /// publish, so it can arrive after the span was superseded, and rolling back the current span
    /// on a stale answer would stop a clock the server never objected to.
    /// </summary>
    public void HandleTrackingConflict(string entryId)
    {
        if (!_tracker.AbandonRunningSpan(entryId))
        {
            return;
        }

        _displayStart = null;
        LiveSyncBlocked = false;
        Notice = "Already tracking on another machine — stop it there first.";
        RaiseTrackingState();
    }

    public void SelectProject(string projectId, string? taskId)
    {
        _displayStart = null;
        Selection = new StoredSelection(projectId, taskId);
        if (_userId is { } userId)
        {
            _selectionStore.Save(Selection, userId);
        }

        // A project switch DOES re-attribute the time, so a running span is closed and reopened
        // under the new selection. (A note change does not — see TimeTracker.SetNote.)
        if (_tracker.State is TrackerState.Tracking)
        {
            _tracker.Stop();
            _tracker.Start(projectId, taskId, NoteOrNull());
            TrackingStarted?.Invoke();
        }
        else if (_tracker.State is TrackerState.Paused)
        {
            _tracker.Reselect(new TimeTracker.Selection(projectId, taskId, NoteOrNull()));
        }

        RaiseTrackingState();
    }

    /// <summary>Restore the sticky selection for this user, dropping anything now stale.</summary>
    public void RestoreSelection(string userId)
    {
        UserId = userId;
        Selection = SelectionResolver.Resolve(_selectionStore.Load(userId), _projects);
    }

    /// <summary>
    /// Sign-out: drop everything user-specific from memory so the next person to sign in on this
    /// machine cannot inherit it (CLAUDE.md §1).
    /// </summary>
    public void Reset()
    {
        IsReady = false;
        UserId = null;
        Projects = [];
        Selection = null;
        Totals = null;
        PendingCount = 0;
        LiveSyncBlocked = false;
        Notice = null;
        Note = string.Empty;
        Query = string.Empty;
        _displayStart = null;
        RaiseTrackingState();
    }

    /// <summary>Called once a second while the popup is open, to advance the live elapsed clock.</summary>
    public void Tick() =>
        Raise(nameof(ElapsedLabel), nameof(Elapsed), nameof(TodayLabel), nameof(WeekLabel), nameof(MonthLabel));

    private string? NoteOrNull() => string.IsNullOrWhiteSpace(_note) ? null : _note;

    private void RaiseTrackingState()
    {
        Raise(
            nameof(IsTracking),
            nameof(IsPaused),
            nameof(CanStart),
            nameof(CanStop),
            nameof(ElapsedLabel),
            nameof(Elapsed),
            nameof(TodayLabel),
            nameof(WeekLabel),
            nameof(MonthLabel));

        // Every path that changes the tracker's state ends here, the auto layer included (through
        // RefreshFromTracker), so this is the one place a stop can be observed.
        var tracking = IsTracking;
        if (_wasTracking && !tracking)
        {
            TrackingStopped?.Invoke();
        }

        _wasTracking = tracking;
    }

    private bool Set<T>(
        ref T field,
        T value,
        string[]? also = null,
        [CallerMemberName] string? propertyName = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value))
        {
            return false;
        }

        field = value;
        Raise(propertyName!);
        if (also is not null)
        {
            Raise(also);
        }

        return true;
    }

    private void Raise(params string[] names)
    {
        foreach (var name in names)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
        }
    }
}
